Wrangler2 release (#3909)
cloudflare/cloudflare-docs: content/workers/wrangler/migration/eject-webpack.md

---
pcx-content-type: how-to
title: Ejecting Webpack
weight: 3
---

> ## Migrating webpack projects from `wrangler` version 1

Reviewer: ```suggestion
## Migrate webpack projects from `wrangler` version 1
```

> Previous versions of `wrangler` offered rudimentary support for [webpack](https://webpack.js.org/) with the `type` and `webpack_config` keys in [`wrangler.toml`](/workers/wrangler/configuration/). Starting with version 2.0,0, `wrangler` no longer supports the `type` and `webpack_config` keys, but you can still use webpack with your Workers.

Reviewer: ```suggestion
Previous versions of `wrangler` offered rudimentary support for [webpack](https://webpack.js.org/) with the `type` and `webpack_config` keys in [`wrangler.toml`](/workers/wrangler/configuration/). Starting with Wrangler 2, `wrangler` no longer supports the `type` and `webpack_config` keys, but you can still use webpack with your Workers.
```

> As a developer using webpack with Workers, you may be one of four categories:

Reviewer: ```suggestion
As a developer using webpack with Workers, you may be in one of four categories:
```

1. [I use `[build]` to run webpack (or another bundler) external to `wrangler`.](#i-use-build-to-run-webpack-or-another-bundler-external-to-wrangler)

> 2. [I use `type = webpack`, but don't provide my own configuration and let `wrangler` take care of it.](#i-use-type--webpack-but-dont-provide-my-own-configuration-and-let-wrangler-take-care-of-it)

Reviewer: ```suggestion
2. [I use `type = webpack`, but do not provide my own configuration and let `wrangler` take care of it.](#i-use-type--webpack-but-dont-provide-my-own-configuration-and-let-wrangler-take-care-of-it)
```

3. [I use `type = webpack` and `webpack_config = <path/to/webpack.config.js>` to handle JSX, Typescript, WebAssembly, HTML files, and other non-standard filetypes.](#i-use-type--webpack-and-webpack_config--pathtowebpackconfigjs-to-handle-jsx-typescript-webassembly-html-files-and-other-non-standard-filetypes)

4. [I use `type = webpack` and `webpack_config = <path/to/webpack.config.js>` to perform code-transforms and/or other code-modifying functionality.](#i-use-type--webpack-and-webpack_config--pathtowebpackconfigjs-to-perform-code-transforms-andor-other-code-modifying-functionality)

If you do not see yourself represented, [file an issue](https://github.com/cloudflare/wrangler2/issues/new/choose) and we can assist you with your specific situation and improve this guide for future readers.

> ### I use `[build]` to run webpack (or another bundler) external to wrangler.

Reviewer: ```suggestion
### I use `[build]` to run webpack (or another bundler) external to Wrangler.
```

`wrangler` 2 supports the `[build]` key, so your Workers will continue to build using your own setup.

> ### I use `type = webpack`, but don't provide my own configuration and let `wrangler` take care of it.

Reviewer: ```suggestion
### I use `type = webpack`, but do not provide my own configuration and let `wrangler` take care of it.
```

`wrangler` will continue to take care of it. Remove `type = webpack` from your `wrangler.toml` file.

### I use `type = webpack` and `webpack_config = <path/to/webpack.config.js>` to handle JSX, Typescript, WebAssembly, HTML files, and other non-standard filetypes.

> As of version 2.0.0, `wrangler` has built-in support for this use case due to the [module system](/workers/wrangler/module-system/).

Reviewer: ```suggestion
As of Wrangler 2, `wrangler` has built-in support for this use case due to the [module system](/workers/wrangler/module-system/).
```

1. Add [wranglerjs-compat-webpack-plugin](TODO: npm link) as a `devDependency`

[wrangler-js](https://www.npmjs.com/package/wrangler-js), shipped as a separate library from [wrangler 1](https://www.npmjs.com/package/@cloudflare/wrangler/v/1.19.11), is a Node script that configures and executes [webpack 4](https://unpkg.com/browse/wrangler-js@0.1.11/package.json) for you. When you set `type = webpack`, wrangler 1 would execute this script for you. We have ported the functionality over to a new package, [wranglerjs-compat-webpack-plugin](TODO: link), which you can use as a [webpack plugin](https://v4.webpack.js.org/configuration/plugins/).

To do that, you will need to add it as a dependency:

```sh
npm install --save-dev webpack@^4.46.0 webpack-cli wranglerjs-compat-webpack-plugin
# or
yarn add --dev webpack@4.46.0 webpack-cli wranglerjs-compat-webpack-plugin
```

You should see this reflected in your `package.json` file:

```json
{
  "name": "my-worker",
  "version": "x.y.z",
  // ...
  "devDependencies": {
    // ...
    "wranglerjs-compat-webpack-plugin": "^x.y.z",
    "webpack": "^4.46.0",
    "webpack-cli": "^x.y.z"
  }
}
```

> 2. Add [wranglerjs-compat-webpack-plugin] to `webpack.config.js`

Reviewer: Is this supposed to be 1. and not 2.?

Modify your `webpack.config.js` file to include the plugin you just installed.

```js
const {
  WranglerJsCompatWebpackPlugin,
} = require("wranglerjs-compat-webpack-plugin");

module.exports = {
  // ...
  plugins: [new WranglerJsCompatWebpackPlugin()],
};
```

2. Add a build script your `package.json`

```json
{
  "name": "my-worker",
  "verion": "2.0.0",
  // ...
  "scripts": {
    "build": "webpack" // <-- Add this line!
    // ...
  }
}
```

3. Remove unsupported entries from your `wrangler.toml`

Remove the `type` and `webpack_config` keys from your `wrangler.toml` file, as they are not supported anymore.

```toml
# Remove these!
type = "webpack"
webpack_config = "webpack.config.js"
```

4. Tell `wrangler` how to bundle your Worker

> `wrangler` no longer has any knowledge of how to build your worker, so you will need to tell it how to call webpack and where to look for webpack's output. This translates into two fields:

Reviewer: ```suggestion
`wrangler` no longer has any knowledge of how to build your Worker. You will need to tell it how to call webpack and where to look for webpack's output. This translates into two fields:
```
